### PR TITLE
Fixes #43: Custom rules for local vhosts now work

### LINKED ISSUE #43: Local vhost custom rules are ignored
When providing a local vhost template, the custom rules are never imported. [Looking at the code](https://github.com/devilbox/vhost-gen/blob/master/bin/vhost-gen#L714), it would seem that is because the template is never checked when importing the custom rules - only the global configuration.

I've created a PR to address this.

### PR DESCRIPTION
**Pull request by CameronHall**
Local vhost custom rules are no longer ignored

### CODE REVIEW

**Comment by cytopia** [comment]
@CameronHall thanks for this PR.

Can you give me some more details why this should not be working? I've just tried it from `master` branch:

**1. Adjusted `etc/conf.yml` in this repo (added to `custom:` section):**

`etc/conf.yml:`
```yml
...
custom:
  test
...
```
Just to show what has changed:
```bash
$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   etc/conf.yml

no changes added to commit (use "git add" and/or "git commit -a")
```
```diff
$ git diff
diff --git a/etc/conf.yml b/etc/conf.yml
index 94f8b75..41139e7 100644
--- a/etc/conf.yml
+++ b/etc/conf.yml
@@ -68,7 +68,7 @@ conf_dir: /etc/nginx/conf.d
 #   custom statement 1
 #   custom statement 2
 custom:
-
+  test

 # Vhost definition
 vhost:
```


**2. Run `vhost-gen` in this repo:**
```shell
./bin/vhost-gen -p /shared -n www.google.de -c etc/conf.yml -t etc/templates/
```

**3. It shows me the custom section being added:**
```nginx
...
    # Status Page
    location ~ /server-status/ {
        stub_status on;
        access_log off;
    }

    # Custom directives
    test
}
```

**Comment by CameronHall** [comment]
Hey @cytopia,

I appreciate you taking the time to reply and apologies for providing no steps to reproduce and a vague description. 

**Background**
I discovered this while using devilbox/docker-apache-2.4 and I have only tested the behaviour with the provided `data/vhost-gen/mass.yml` config. I spent a bit of time guessing my way around how this worked exactly, so I won't be surprised if I've just stuffed something up along the way 😄 

**Reproduction**
I've created a short script to reproduce the issue, if you run it from within your vhost-gen project root you should get the same output as me.

```
#!/bin/bash
mkdir tests/resources && curl https://raw.githubusercontent.com/devilbox/docker-apache-2.4/master/data/vhost-gen/mass.yml > tests/resources/mass.yml
echo 'custom: ServerAlias potato.com' > tests/resources/apache24.yml
bin/vhost-gen -n dog.com -c tests/resources/mass.yml -o tests/resources -p /web -t etc/templates
```

**Output before changes**
```xml
    ...
    # Deny Definition
    <FilesMatch "/\.ht.*">
        Order allow,deny
        Deny from all
    </FilesMatch>


    # Custom directives

</VirtualHost>
```

**Output after changes**
```xml
    ...
    # Deny Definition
    <FilesMatch "/\.ht.*">
        Order allow,deny
        Deny from all
    </FilesMatch>


    # Custom directives
    ServerAlias potato.com
</VirtualHost>
```

**Comment by cytopia** [comment]
I can reproduce this behaviour. Currently still trying to figure out why my example works and yours doesn't

**Comment by science695** [comment]
@cytopia I'm not super familiar with yaml syntax, but your example has the custom text on the next line down, while his example is inline.

Could that be the issue?

**Comment by cytopia** [comment]
> @cytopia I'm not super familiar with yaml syntax, but your example has the custom text on the next line down, while his example is inline.
> 
> Could that be the issue?

Changed it to:
```yaml
custom: |
  test
```
and
```yaml
custom: test
```
which works for me.

The issue is rather with `-o` (override) as seen in @CameronHall example

**Comment by cytopia** [comment]
**FYI-Update:**  @CameronHall 

Currently adding full debug output to `vhost-gen`. It already seems, that your custom definition should go into the configuration file, not the template:


```bash
$ bin/vhost-gen -n dog.com -c tests/resources/mass.yml -o tests/resources -p /web -t etc/templates -vv >/dev/null
vhost-gen: [INFO]  (2020-12-05 10:21:40): Loading configuration file        (-c): tests/resources/mass.yml
vhost-gen: [INFO]  (2020-12-05 10:21:40): Loading vhost template (global)   (-t): etc/templates/apache24.yml
vhost-gen: [INFO]  (2020-12-05 10:21:40): Loading vhost template (override) (-o): tests/resources/apache24.yml
vhost-gen: [INFO]  (2020-12-05 10:21:40): Creating vhost type: http-only (plain)
vhost-gen: [DEBUG] (2020-12-05 10:21:40): [status] Retrieving cfg 'enable': False
vhost-gen: [DEBUG] (2020-12-05 10:21:40): [status] Returning:
vhost-gen: [DEBUG] (2020-12-05 10:21:40): [custom] Retrieving cfg 'custom':
vhost-gen: [DEBUG] (2020-12-05 10:21:40): [custom] Returning': None
vhost-gen: [INFO]  (2020-12-05 10:21:40): Using vhost name: dog.com__TLD__
```

The templates should not be modified, as they already contain all required server-specific configuration sections. The config file itself is used to replace values into the templates.
In your case, you are trying to hardcode values in the template. this should however go into `tests/resources/mass.yml` instead.

You've probably mixed up template and config?

**Comment by cytopia** [comment]
@CameronHall have a look at this version (not complete yet, but with `-vvv` you can see what exactly it is doing)

https://github.com/devilbox/vhost-gen/pull/45

**Comment by CameronHall** [comment]
@cytopia my specific use case for this was to add a `ServerAlias` to virtual hosts for two independent projects in [devilbox/docker-apache-2.4](https://github.com/devilbox/docker-apache-2.4) where the aliases vary between projects. The command I provided is a slight variation of [this](https://github.com/devilbox/docker-apache-2.4/blob/master/data/create-vhost.sh#L33).
 
I understand what you're saying with respect to hardcoding values, but I think the override should be considered "vhost specific configuration"? Instead of a "template override". That's just my opinion for this specific use case, I don't any exposure to using this tool outside of my own use cases - so I'd appreciate your perspective 😄

File: bin/vhost-gen [hunk 1]
```diff
@@ -849,11 +854,11 @@ def get_vhost_ssl(
             "__INDEX__": vhost_get_index(config, verbose),
             "__ACCESS_LOG__": vhost_get_access_log(config, server_name + "_ssl"),
             "__ERROR_LOG__": vhost_get_error_log(config, server_name + "_ssl"),
-            "__PHP_FPM__": str_indent(vhost_get_php_fpm(config, tpl, docroot, proxy, verbose), 4),
-            "__ALIASES__": str_indent(vhost_get_aliases(config, tpl, verbose), 4),
-            "__DENIES__": str_indent(vhost_get_denies(config, tpl, verbose), 4),
-            "__SERVER_STATUS__": str_indent(vhost_get_server_status(config, tpl, verbose), 4),
-            "__CUSTOM__": str_indent(vhost_get_custom_section(config, verbose), 4),
+            "__PHP_FPM__": str_indent(vhost_get_php_fpm(config, tpl, docroot, proxy), 4),
```
Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
            "__PHP_FPM__": str_indent(vhost_get_php_fpm(config, tpl, docroot, proxy, verbose), 4),
```

File: bin/vhost-gen [hunk 2]
```diff
@@ -849,11 +854,11 @@ def get_vhost_ssl(
...
+            "__ALIASES__": str_indent(vhost_get_aliases(config, tpl), 4),
```
Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
            "__ALIASES__": str_indent(vhost_get_aliases(config, tpl, verbose), 4),
```

File: bin/vhost-gen [hunk 3]
```diff
@@ -849,11 +854,11 @@ def get_vhost_ssl(
...
+            "__DENIES__": str_indent(vhost_get_denies(config, tpl), 4),
```
Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
            "__DENIES__": str_indent(vhost_get_denies(config, tpl, verbose), 4),
```

File: bin/vhost-gen [hunk 4]
```diff
@@ -849,11 +854,11 @@ def get_vhost_ssl(
...
+            "__SERVER_STATUS__": str_indent(vhost_get_server_status(config, tpl), 4),
```
Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
            "__SERVER_STATUS__": str_indent(vhost_get_server_status(config, tpl, verbose), 4),
```

File: bin/vhost-gen [hunk 5]
```diff
@@ -849,11 +854,11 @@ def get_vhost_ssl(
...
+            "__CUSTOM__": str_indent(vhost_get_custom_section(config, tpl), 4),
```
Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
            "__CUSTOM__": str_indent(vhost_get_custom_section(config, tpl, verbose), 4),
```

File: bin/vhost-gen [hunk 6]
```diff
@@ -766,12 +766,17 @@ def vhost_get_server_status(config, template, verbose):
...
 
-def vhost_get_custom_section(config, verbose):
-    # type: (Dict[str, Any], int) -> str
+def vhost_get_custom_section(config, template, verbose):
     """Get virtual host custom directives."""
     output = to_str(config["custom"])
+    if output:
+        output += "\n"
     log(3, "[custom]  Retrieving cfg 'custom': {}".format(output), verbose)
     log(3, "[custom]  Returning': {}".format(output), verbose)
```
Comment (cytopia) [inline bin/vhost-gen]: This should be right above the return statemt to see the final return value

**Review by cytopia — CHANGES_REQUESTED** [review]

File: bin/vhost-gen [hunk 7]
```diff
@@ -766,12 +766,17 @@ def vhost_get_server_status(config, template, verbose):
     return output
 
 
-def vhost_get_custom_section(config, verbose):
-    # type: (Dict[str, Any], int) -> str
+def vhost_get_custom_section(config, template, verbose):
```
Comment (cytopia) [inline bin/vhost-gen:769]:
```suggestion
def vhost_get_custom_section(config, template, verbose):
    # type: (Dict[str, Any], Dict[str, Any], int) -> str
```

File: bin/vhost-gen [hunk 8]
```diff
@@ -766,12 +766,19 @@ def vhost_get_server_status(config, template, verbose):
...
+    # type: (Dict[str, Any], Dict[str, Any], int) -> str
     """Get virtual host custom directives."""
     output = to_str(config["custom"])
+    if output:
+        output += "\n"
+
+    if "custom" in template:
+        output += to_str(template["custom"])
+
     log(3, "[custom]  Retrieving cfg 'custom': {}".format(output), verbose)
```
Comment (cytopia) [inline bin/vhost-gen]: This was one too much. The first log statement is about what has been retrieved from the config, so this should be directory below the first `output = to_str(config["custom"])` line

Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
```

File: bin/vhost-gen [hunk 9]
```diff
@@ -766,12 +766,19 @@ def vhost_get_server_status(config, template, verbose):
     return output
 
 
-def vhost_get_custom_section(config, verbose):
-    # type: (Dict[str, Any], int) -> str
+def vhost_get_custom_section(config, template, verbose):
+    # type: (Dict[str, Any], Dict[str, Any], int) -> str
     """Get virtual host custom directives."""
     output = to_str(config["custom"])
```
Comment (cytopia) [inline bin/vhost-gen:772]:
```suggestion
    output = to_str(config["custom"])
    log(3, "[custom]  Retrieving cfg 'custom': {}".format(output), verbose)
```

File: bin/vhost-gen [hunk 10]
```diff
@@ -766,12 +766,19 @@ def vhost_get_server_status(config, template, verbose):
...
+    if output:
+        output += "\n"
+
+    if "custom" in template:
+        output += to_str(template["custom"])
+
     log(3, "[custom]  Retrieving cfg 'custom': {}".format(output), verbose)
     log(3, "[custom]  Returning': {}".format(output), verbose)
+
```
Comment (cytopia) [inline bin/vhost-gen]:
```suggestion
```

**Review by cytopia — CHANGES_REQUESTED** [review]

**Review by cytopia — COMMENTED** [review]
Can you bump the version to:

`bin/vhost-gen`
```diff
- VERSION = "1.0.2"
+ VERSION = "1.0.3"
```

`setup.py`
```diff
-    version="1.0.2",
+    version="1.0.3",
```

**Comment by CameronHall** [comment]
The version number has been bumped

**Review by cytopia — APPROVED** [review]
Doesn't hurt to have this feature :-) Thanks for the contribution :+1:

**Comment by cytopia** [comment]
Available via pip now: https://pypi.org/project/vhost-gen/

**Pull request merged**